### PR TITLE
ODSC-46642/pass model by reference

### PR DESCRIPTION
**Pull request by mingkang111**
# Description
https://jira.oci.oraclecorp.com/browse/ODSC-46642
Improve the [LargeArtifactUploader](https://github.com/oracle/accelerated-data-science/blob/main/ads/model/artifact_uploader.py#L106) class to understand OCI paths to upload model artifacts to the model catalog by reference.

# Changed
* Case1: user want to copy the model artifact in [ARTIFACT_PATH](https://github.com/oracle/accelerated-data-science/blob/main/ads/model/datascience_model.py#L131) to [bucket_uri](https://github.com/oracle/accelerated-data-science/blob/main/ads/model/datascience_model.py#L547), and then upload to MC. (this is the current case we are supporting).
* Case 2: user wants to upload model artifacts in [bucket_uri](https://github.com/oracle/accelerated-data-science/blob/main/ads/model/datascience_model.py#L547) to MC. (this is the new case we want to support)

Allow the [ARTIFACT_PATH](https://github.com/oracle/accelerated-data-science/blob/main/ads/model/datascience_model.py#L131)  to be oci:// path, and only allow for model zip file. e.g: `oci://xxx@tenency/my_model.zip` . Then if user is targeting on case2,  user can pass `ARTIFACT_PATH="oci://xxx@tenency/my_model.zip"`, we can set [bucket_uri](https://github.com/oracle/accelerated-data-science/blob/main/ads/model/datascience_model.py#L547)  for them internally.

# User experience
* Uploading any size model (include the LLM models) already exists in Object Storage
```
import ads
from ads.model import DataScienceModel

ads.set_auth("resource_principal")

MODEL_DISPLAY_NAME = "my_large_model"
OCI_BUCKET = "oci://bucket@namespace/prefix/my_large_model.zip"

model = (DataScienceModel()
         .with_display_name(MODEL_DISPLAY_NAME)
         .with_artifact(OCI_BUCKET)
         .create()
        ) 
```

* Uploading the large models not exist in Object Storage, but in local
```
import ads
from ads.model import DataScienceModel

ads.set_auth("resource_principal")

MODEL_DISPLAY_NAME = "my_large_model"
ARTFICAT_FILE_NAME = "./my_large_model.zip"  
OCI_BUCKET = "oci://bucket@namespace/prefix/my_large_model.zip"

model = (DataScienceModel()
         .with_display_name(MODEL_DISPLAY_NAME)
         .with_artifact(ARTFICAT_FILE_NAME)
         .create(
            bucket_uri=OCI_BUCKET, # model artifact already in this path
            overwrite_existing_artifact=True, # Overwrite target bucket artifact if exists.
            remove_existing_artifact=True # Whether artifacts uploaded to the object storage bucket need to be removed or not.
         )
        ) 
```
* Uploading the small model in local
```
import ads
from ads.model import DataScienceModel

ads.set_auth("resource_principal")

MODEL_DISPLAY_NAME = "my_large_model"
ARTFICAT_FILE_NAME = "./my_large_model.zip"  

model = (DataScienceModel()
         .with_display_name(MODEL_DISPLAY_NAME)
         .with_artifact(ARTFICAT_FILE_NAME)
         .create( )
        ) 
```

# Test
* Unit test 
Screenshot 2023-09-21 at 4 44 10 PM

* Test create MC for the model in OS
Screenshot 2023-09-21 at 5 48 29 PM

Screenshot 2023-09-21 at 5 51 35 PM

### CODE REVIEW

File: ads/model/artifact_uploader.py [hunk 1]
```diff
@@ -179,11 +190,20 @@ def __init__(
         parallel_process_count: (int, optional).
             The number of worker processes to use in parallel for uploading individual parts of a multipart upload.
         """
+        self.auth = auth or dsc_model.auth
+        if ObjectStorageDetails.is_oci_path(artifact_path):
+            if not artifact_path.endswith(".zip"):
+                raise ValueError(
+                    f"The `artifact_path={artifact_path}` is invalid. Please check APIs doc to see the possible values for it."
```
Comment (mrDzurb) [inline ads/model/artifact_uploader.py:197]: Let's make the message more informative for user. Something similar that you've put o the method description.

**Comment by mrDzurb** [comment]
If user provides the path to zip in OS bucket, let's auto extract the 
```bucket_uri=OCI_BUCKET, # model artifact already in this path``` 
However if user provides different ```bucket_uri```, then let's show a warning and say that this argument will be ignored for the case when ```.with_artifact(ARTFICAT_FILE_NAME)``` is a path to zip archive. Also let's always use the large artifact uploader if user provides the path to zip archive in Object Storage.

File: ads/model/artifact_uploader.py (hunk 1 above)
Comment (mingkang111) [reply ads/model/artifact_uploader.py]: Fixed!

**Comment by mingkang111** [comment]
> If user provides the path to zip in OS bucket, let's auto extract the `bucket_uri=OCI_BUCKET, # model artifact already in this path` However if user provides different `bucket_uri`, then let's show a warning and say that this argument will be ignored for the case when `.with_artifact(ARTFICAT_FILE_NAME)` is a path to zip archive. Also let's always use the large artifact uploader if user provides the path to zip archive in Object Storage.

Yes. As we discussed, the warning messages have been added in `ads/model/datascience_model.py`.  When the user passes a remote path for the model artifact, we will automatically assign value to `bucket_uri`, so the `LargeArtifactUploader` will always be used. Thanks for the feedback!

**Review by mrDzurb — APPROVED** [review]

File: ads/model/datascience_model.py [hunk 2]
```diff
@@ -668,6 +673,13 @@ def upload_artifact(
                 "timeout": timeout,
             }
 
+        if ObjectStorageDetails.is_oci_path(self.artifact):
+            if bucket_uri and bucket_uri != self.artifac:
```
Comment (z7ye) [inline ads/model/datascience_model.py:677]: self.artifac -> self.artifact

**Review by z7ye — CHANGES_REQUESTED** [review]

File: ads/model/datascience_model.py (hunk 2 above)
Comment (mingkang111) [reply ads/model/datascience_model.py]: Thanks @z7ye ! Fixed!

**Review by z7ye — APPROVED** [review]

**Review by mrDzurb — APPROVED** [review]

**Pull request merged**